### PR TITLE
Allow parties to be hosted

### PR DESCRIPTION
**Pull request by roastario**
See: https://github.com/corda/accounts/pull/2 and https://github.com/corda/accounts-demo/pull/3 for usages of this new API.

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/keys/BasicHSMKeyManagementService.kt [hunk 1]
```diff
@@ -84,7 +83,7 @@ class BasicHSMKeyManagementService(cacheFactory: NamedCacheFactory, val identity
     }
 
     override fun filterMyKeys(candidateKeys: Iterable<PublicKey>): Iterable<PublicKey> = database.transaction {
-        identityService.stripNotOurKeys(candidateKeys)
+        (identityService.stripNotOurKeys(candidateKeys) + candidateKeys.filter { containsPublicKey(it) }).toSet()
```
Comment (kasiastreich) [inline node/src/main/kotlin/net/corda/node/services/keys/BasicHSMKeyManagementService.kt]: Maybe you can also explain here why this change is needed. On this branch you don't record anything into the originalKeyMap/keyMap. Is it a faulty implementation of BasicHSMKeyManagementService that it takes into account only keys that are in identityService?

File: build.gradle [hunk 2]
```diff
@@ -4,7 +4,7 @@ buildscript {
     file("$projectDir/constants.properties").withInputStream { constants.load(it) }
 
     // Our version: bump this on release.
-    ext.corda_release_version = constants.getProperty("cordaVersion")
+    ext.corda_release_version = "4.1"
```
[the remark this answers is not in the document]
Comment (roastario) [reply build.gradle]: just for my maven repo to not overwrite "proper" corda. this will be reverted if we seem like getting approval

File: core/src/main/kotlin/net/corda/core/identity/HostedParty.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import java.security.PublicKey
+
+@CordaSerializable
+class HostedParty(val host: AbstractParty, owningKey: PublicKey) : AbstractParty(owningKey) {
+    override fun nameOrNull(): CordaX500Name? {
+        return host.nameOrNull();
+    }
+
+    override fun ref(bytes: OpaqueBytes): PartyAndReference {
+        return PartyAndReference(this, OpaqueBytes(owningKey.encoded))
```
[the remark this answers is not in the document]
Comment (roastario) [reply core/src/main/kotlin/net/corda/core/identity/HostedParty.kt]: true story. 

fixing.

File: core/src/main/kotlin/net/corda/core/identity/HostedParty.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+import net.corda.core.contracts.PartyAndReference
+import net.corda.core.serialization.CordaSerializable
+import net.corda.core.utilities.OpaqueBytes
+import java.security.PublicKey
+
+@CordaSerializable
+class HostedParty(val host: AbstractParty, owningKey: PublicKey) : AbstractParty(owningKey) {
+    override fun nameOrNull(): CordaX500Name? {
+        return host.nameOrNull();
```
[the remark this answers is not in the document]
Comment (roastario) [reply core/src/main/kotlin/net/corda/core/identity/HostedParty.kt]: java ¬_¬

File: core/src/main/kotlin/net/corda/core/identity/AbstractParty.kt [hunk 5]
```diff
@@ -30,4 +30,6 @@ abstract class AbstractParty(val owningKey: PublicKey) {
      * ledger.
      */
     fun ref(vararg bytes: Byte) = ref(OpaqueBytes.of(*bytes))
+
+    abstract fun host(): AbstractParty
```
[the remark this answers is not in the document]
Comment (roastario) [reply core/src/main/kotlin/net/corda/core/identity/AbstractParty.kt]: I have changed it to a field with default getter in `AbstractParty`. 

Will do some thinking too about how is possible to do with switch statement.
